pkg/framework/kind/load.go

```diff
@@ -0,0 +1,195 @@
+// variants, and attempts to extract its metadata into the relevant typed struct.
+func BindSomeKind(v cue.Value) (SomeKindMeta, error) {
+	if !v.Exists() {
+		return nil, ErrValueNotExist
+	}
+
+	if meta, err := BindKind[RawMeta](v); err == nil {
+		return meta, nil
+		// return metaToSome(meta), nil
+	}
+	if meta, err := BindKind[CoreStructuredMeta](v); err == nil {
+		return meta, nil
+		// return metaToSome(meta), nil
+	}
+	if meta, err := BindKind[CustomStructuredMeta](v); err == nil {
+		return meta, nil
+		// return metaToSome(meta), nil
+	}
+	return nil, ErrValueNotAKind
+}
+
+func BindKind[T KindMetas](v cue.Value) (T, error) {
+	meta := new(T)
+	if !v.Exists() {
+		return *meta, ErrValueNotExist
+	}
+
+	fw := CUEFramework(v.Context())
+	var kdef cue.Value
+
+	anymeta := any(*meta).(SomeKindMeta)
+	switch anymeta.(type) {
+	case RawMeta:
+		kdef = fw.LookupPath(cue.MakePath(cue.Def("Raw")))
+	case CoreStructuredMeta:
+		kdef = fw.LookupPath(cue.MakePath(cue.Def("CoreStructured")))
+	case CustomStructuredMeta:
+		kdef = fw.LookupPath(cue.MakePath(cue.Def("CustomStructured")))
+	default:
+		// unreachable so long as all the possibilities in KindMetas have switch branches
+		panic("unreachable")
+	}
+
+	item := v.Unify(kdef)
+	// TODO recall and comment on the difference between Err and Validate
+	if item.Err() != nil {
+		return *meta, ewrap(item.Err(), ErrValueNotAKind)
+	}
+	if err := item.Validate(); err != nil {
+		return *meta, ewrap(item.Err(), ErrValueNotAKind)
+	}
+	if err := item.Decode(meta); err != nil {
+		// Should only be reachable if CUE and Go framework types have diverged
+		panic(errors.Details(err, nil))
+	}
+
+	return *meta, nil
+}
+
+// ParsedKindDefinition is the result of a successful call to ParseKindFS. It
+// represents a kind definition in partially-processed form, amenable for
+// further processing into
+type ParsedKindDefinition struct {
+	// V is the cue.Value corresponding to the entire Kind declaration.
+	V cue.Value
+	// Meta configuration for the kind.
+	Meta SomeKindMeta
+}
+
+// BindKindLineage binds the lineage for the parsed kind. nil, nil is returned for raw kinds.
+//
+// For kinds with a corresponding Go type, it left to the caller to associate that Go type
+// with the returned lineage by calling [thema.BindType].
+func (pk *ParsedKindDefinition) BindKindLineage(rt *thema.Runtime, opts ...thema.BindOption) (thema.Lineage, error) {
+	switch pk.Meta.(type) {
+	case RawMeta:
+		return nil, nil
+	case CoreStructuredMeta, CustomStructuredMeta:
+		return thema.BindLineage(pk.V.LookupPath(cue.MakePath(cue.Str("lineage"))), rt, opts...)
+	default:
+		panic("unreachable")
+	}
+}
+
+// ParseKindFS takes an fs.FS and validates that it represents one
+// of the valid kind variants. On success, it returns a representation
+// of the entire kind definition contained in the provided kfs.
+func ParseKindFS(kfs fs.FS, rt *thema.Runtime) (*ParsedKindDefinition, error) {
+	panic("TODO")
+}
```

Comment:
We already have a similar function for plugins - `pkg/plugins/pfs.ParsePluginFS`. The idea behind a func at this particular level of abstraction receiving an `fs.FS` is that we can use it:

* **at codegen-time,** when we scan the real filesystem and construct an `os.DirFS` - like [here](https://github.com/grafana/grafana/blob/main/public/app/plugins/gen.go#L60)) _and_ at runtime when loading up things in our `embed.FS` - like here
* **at runtime,** when we [load the same stuff](https://github.com/grafana/grafana/blob/f9e426cf80b83dac6abf2a22491795630deba31f/pkg/plugins/pfs/corelist/loadlist_gen.go#L35), but this time from an `embed.FS`

This is a hard-won lesson about effective code generator design. Requiring that these two different cases be treated symmetrically is a powerful design constraint: it reduces bugs and code duplication, as well as enforcing that we have a clean concept of a loader.